### PR TITLE
feat(shared): implement debt math helpers (accrual, health factor, bo…

### LINKED ISSUE #615: feat(shared): implement protocol math for debt, HF, and accrual
## 📝 Description

The shared crate already compiles. Types, constants, and function signatures live in `contracts/shared`. The math functions still return `SharedError::NotImplemented`.

This issue fills in those bodies so vault and pool can call them. Do not change the signatures. Do not implement borrow, repay, or vault behavior.

Reference: `docs/arch.md` §6.2 (accrual formula) and §10 (health factor).

## ✅ Requirements

- Implement the frozen signatures listed below. Leave names and parameters unchanged.
- Accrual: `principal * rate_bps / 10_000 * elapsed_seconds / 31_536_000` with checked arithmetic. No `unwrap` on overflow.
- `health_factor_bps`: floor division. `debt == 0` returns `i128::MAX`.
- `borrow_limit_from_collateral`: floor `collateral_value * max_ltv_bps / 10_000`.
- `ceil_div`: round away from zero toward +inf for positive values.
- `Debt::total`: checked `principal + accrued_interest`.
- Keep the crate `no_std`.
- Add tests in `contracts/shared/src/tests.rs` (create the file and `mod tests;` in `lib.rs`).

## 🎯 Acceptance Criteria

- [ ] No remaining `NotImplemented` in the functions listed below.
- [ ] `cargo test -p shared --all-features` is green.
- [ ] Signatures are unchanged.
- [ ] Screenshots of the implemented bodies and of the test run are attached.

## 📁 Expected files to change/structure

- `contracts/shared/src/types.rs` *(implement bodies)*
- `contracts/shared/src/lib.rs` *(add `mod tests`)*
- `contracts/shared/src/tests.rs` *(create)*

### Where to implement

Signatures are already declared. Implement the **body** of each:

| Function | File |
| --- | --- |
| `Debt::total` | `contracts/shared/src/types.rs` |
| `accrue_linear_interest` | `contracts/shared/src/types.rs` |
| `health_factor_bps` | `contracts/shared/src/types.rs` |
| `borrow_limit_from_collateral` | `contracts/shared/src/types.rs` |
| `ceil_div` | `contracts/shared/src/types.rs` |

Constants are already in `contracts/shared/src/constant.rs`. Do not change them.

### Mandatory tests

File: `contracts/shared/src/tests.rs`

```rust
#[test]
fn test_accrue_linear_interest_one_year_eight_percent();
// principal=1_000_000_000, rate=800, elapsed=31_536_000 => interest=8_000_000

#[test]
fn test_accrue_linear_interest_zero_elapsed_is_zero();

#[test]
fn test_accrue_linear_interest_rejects_negative_principal();

#[test]
fn test_debt_total_is_principal_plus_accrued();

#[test]
fn test_health_factor_bps_example_from_arch();
// collateral=10_000, LT=8000, debt=7_500 => HF_bps = 10666

#[test]
fn test_health_factor_zero_debt_is_max();

#[test]
fn test_borrow_limit_seventy_percent();
// collateral=10_000_000, ltv=7000 => 7_000_000

#[test]
fn test_ceil_div_rounds_up();
```

```bash
cargo test -p shared --all-features
```

### Screenshots required on the PR

- Screenshot of `accrue_linear_interest` and `health_factor_bps` bodies in `types.rs`.
- Screenshot of `test_accrue_linear_interest_one_year_eight_percent ... ok`.

### PR DESCRIPTION
**Pull request by yunus-dev-codecrafter**
closes #615 
…rrow limit, ceil div) with tests

# Implement shared debt math helpers

## Summary

Fills in the bodies of the frozen math functions in `contracts/shared` so `vault` and `pool` can call them. No signatures were changed; no borrow, repay, or vault behavior was implemented.

## Changes

- `contracts/shared/src/types.rs` — implemented:
  - `Debt::total` — checked `principal + accrued_interest`, returns `Overflow` on overflow.
  - `accrue_linear_interest` — `principal * rate_bps / 10_000 * elapsed_seconds / 31_536_000` with checked arithmetic (no unwrap), rejects negative principal with `InvalidAmount`.
  - `health_factor_bps` — floor division of `collateral_value * liquidation_threshold_bps / debt`; returns `i128::MAX` when `debt == 0`.
  - `borrow_limit_from_collateral` — floor `collateral_value * max_ltv_bps / 10_000`.
  - `ceil_div` — rounds toward +inf (away from zero) for positive values.
- `contracts/shared/src/lib.rs` — added `#[cfg(test)] mod tests;`.
- `contracts/shared/src/tests.rs` — new file with 8 unit tests covering the above, per the issue's mandatory test list.

## Test results

`cargo test -p shared --all-features`:

```
running 8 tests
test tests::test_accrue_linear_interest_one_year_eight_percent ... ok
test tests::test_accrue_linear_interest_rejects_negative_principal ... ok
test tests::test_accrue_linear_interest_zero_elapsed_is_zero ... ok
test tests::test_borrow_limit_seventy_percent ... ok
test tests::test_ceil_div_rounds_up ... ok
test tests::test_debt_total_is_principal_plus_accrued ... ok
test tests::test_health_factor_bps_example_from_arch ... ok
test tests::test_health_factor_zero_debt_is_max ... ok

test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

## Notes

- The crate remains `no_std`.
- One discrepancy in the issue's expected value was resolved: the mandatory accrual test comment stated `interest=8_000_000`, but the required formula produces `80_000_000` for `principal=1_000_000_000, rate=800, elapsed=31_536_000`. The test asserts the formula-correct value `80_000_000`.

## Summary by CodeRabbit

* **New Features**
  * Added accurate calculations for debt totals, interest accrual, health factors, collateral-based borrowing limits, and ceiling division.
  * Added validation and overflow handling for edge cases, including zero, negative, rounding, and maximum-value inputs.

* **Tests**
  * Added comprehensive unit tests covering core financial calculations and boundary conditions.

### CODE REVIEW

**Comment by ryzen-xp** [comment]
@abufauzahn  fix your code why failed CI

**Comment by yunus-dev-codecrafter** [comment]
CI fixed, all green

File: contracts/shared/src/types.rs [hunk 1]
```diff
@@ -40,28 +42,60 @@ pub fn accrue_linear_interest(
...
+    }
+    let annual = principal
+        .checked_mul(rate_bps as i128)
+        .ok_or(SharedError::Overflow)?
+        / BPS_DENOMINATOR;
+    annual
+        .checked_mul(elapsed_seconds as i128)
+        .ok_or(SharedError::Overflow)?
+        .checked_div(SECONDS_PER_YEAR as i128)
+        .ok_or(SharedError::DivisionByZero)
```
[the remark this answers is not in the document]
Comment (ryzen-xp) [reply contracts/shared/src/types.rs:56]: @abufauzahn    Fix this plz !!

**Comment by ryzen-xp** [comment]
Hello @abufauzahn  , Nice Job Bro !!  You only needed some fixes that mention above . When U done Then I merged It .

File: contracts/shared/src/types.rs (hunk 1 above)
Comment (ryzen-xp) [inline contracts/shared/src/types.rs:56]: as coderebbitai mention please follow that

**Review by ryzen-xp — CHANGES_REQUESTED** [review]
LGTM !!

**Comment by yunus-dev-codecrafter** [comment]
Done Sir!

**Review by ryzen-xp — APPROVED** [review]
